### PR TITLE
Check matrix processing time

### PR DESCRIPTION
**Pull request by palango**
## Description

Fixes: #5445

- Check matrix response processing time, throw exception if it takes too long
- Update the sync token only after the response has been processed

### Notes

Having dicts as the config options is a pain in the ass, we should work on https://github.com/raiden-network/raiden/issues/4069 after the release.

## PR review check list

Quality check list that cannot be automatically verified.

- Safety
    - [ ] The changes respect the necessary conditions for safety (https://raiden-network-specification.readthedocs.io/en/latest/smart_contracts.html#protocol-values-constraints)
-  Code quality
    - [ ] Error conditions are handled
    - [ ] Exceptions are propagated to the correct parent greenlet
    - [ ] Exceptions are correctly classified as recoverable or unrecoverable
- Compatibility
    - [ ] State changes are forward compatible
    - [ ] Transport messages are backwards and forward compatible
- Commits
    - [ ] Have good messages
    - [ ] Squashed unecessary commits
- If it's a bug fix:
    - Regression test for the bug
        - [ ] Properly covers the bug
        - [ ] If an integration test is used, it could not be written as a unit test
- Documentation
    - [ ] A new CLI flag was introduced, is there documentation that explains usage?
- Specs
    - [ ] If this is a protocol change, are the specs updated accordingly? If so, please link PR from the specs repo.
- Is it a user facing feature/bug fix?
    - [ ] Changelog entry has been added

### CODE REVIEW

**Comment by palango** [comment]
> Looks good so far, but why handle this only in development mode and ignore it in production mode?

Because we never want to crash in production mode

**Comment by palango** [comment]
Mh, these tests pass locally for me.

@hackaugusto thinking about Jochens comment above. If I remember correctly we throw exceptions in other cases and then decide (somewhere?) to log them (on production mode) or crash (on development). Why can't we do this here? (it should be thrown after the `sync` token has been updated)

**Comment by hackaugusto** [comment]
> Why can't we do this here?

We can't really kill the greenlet which is doing the `sync`, there is no code in place to restart it (the assumption is that the task will never die).

If you do want to take this approach, we need to make sure the state of the `sync` greenlet is not lost and put code in place to start it again. However, that is extra work.

File: raiden/exceptions.py [hunk 1]
```diff
@@ -333,3 +333,7 @@ class MintFailed(RaidenError):
 
 class SerializationError(RaidenError):
     """ Invalid data are to be (de-)serialized. """
+
+
+class MatrixProcessingTakingTooLongError(RaidenRecoverableError):
```
Comment (rakanalh) [inline raiden/exceptions.py]:
```suggestion
class MatrixSyncMaxTimeoutReached(RaidenRecoverableError):
```

File: raiden/network/transport/matrix/client.py [hunk 2]
```diff
@@ -475,7 +481,23 @@ def _sync(self, timeout_ms: int = 30000) -> None:
                 len(room["account_data"]["events"]) for room in response["rooms"]["join"].values()
             ),
         )
-        self._handle_response(response, is_first_sync)
+        before_processing = time.time()
+        self._handle_response(response=response, first_sync=is_first_sync)
+        processing_time_s = time.time() - before_processing
```
Comment (rakanalh) [inline raiden/network/transport/matrix/client.py]:
```suggestion
        processing_time_in_seconds = time.time() - before_processing
```

File: raiden/network/transport/matrix/client.py [hunk 3]
```diff
@@ -475,7 +481,23 @@ def _sync(self, timeout_ms: int = 30000) -> None:
...
+
+        # If processing the matrix response takes longer than the poll timout, it cannot be
+        # ensured that all messages have been processed.
+        # Therefore an exception is thrown when in development mode.
+        timeout_s = timeout_ms // 1_000
```
Comment (rakanalh) [inline raiden/network/transport/matrix/client.py]: why don't we just compare the millisecond value?

Comment (palango) [reply raiden/network/transport/matrix/client.py]: because `time()` returns seconds and there is only ´time_ns()`, which makes less sense in my opinion.

**Pull request merged**